fastestimator/backend/get_gradient.py

```diff
@@ -0,0 +1,49 @@
+# Copyright 2019 The FastEstimator Authors. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+from typing import Iterable, Optional, Union
+
+import tensorflow as tf
+import torch
+
+from fastestimator.util.util import NonContext
+
+
+def get_gradient(target: Union[tf.Tensor, torch.Tensor],
+                 sources: Union[Iterable[Union[tf.Tensor, torch.Tensor]], tf.Tensor, torch.Tensor],
+                 higher_order: bool = False,
+                 tape: Optional[tf.GradientTape] = None,
+                 retain_graph: bool = True) -> Union[Iterable[Union[tf.Tensor, torch.Tensor]], tf.Tensor, torch.Tensor]:
+    """calculate gradients of target w.r.t sources
+
+    Args:
+        target: target tensor
+        sources : sequence of source tensors
+        higher_order : whether the gradient will be used for higher order gradients. Defaults to False.
+        tape : tensorflow gradient tape, only needed when using tensorflow backend. Defaults to None.
+        retain_graph : whether to retain pytorch graph, only valid under pytorch backend. Defaults to True.
+
+    Returns:
+        gradients as sequence of tensors
+    """
+    if tape:
+        with NonContext() if higher_order else tape.stop_recording():
+            gradients = tape.gradient(target, sources)
+    else:
+        gradients = torch.autograd.grad(target,
+                                        sources,
+                                        retain_graph=retain_graph,
+                                        create_graph=higher_order,
```

Comment:
not sure that create_graph = True will result in higher order derivatives. When I run two calls with that flag set the result is the same each time

Reply:
it should be the same results, but when create_graph=true, the gradient output will have a .grad field - that's what we are looking for when doing higher order

Reply:
when i run an example of this the result tensor .grad is not populated either